File: LeetCode75_LevelTwo/Path_Sum_III/Path_Sum_III.py
# 437. Path Sum III Python
import collections
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def pathSum(self, root, targetSum):
        """
        :type root: TreeNode
        :type targetSum: int
        :rtype: int
        """
        sums = collections.defaultdict(int)
        sums[0] = 1

        def dfs(root, total):
            count = 0
            if root:
                total += root.val
                count = sums[total - targetSum]

                sums[total] += 1

                count += dfs(root.left, total) + dfs(root.right, total)
                sums[total] -= 1
            return count
        return dfs(root, 0)

File: LeetCode75_LevelTwo/Path_Sum_III/Path_Sum_III.cc
// 437. Path Sum III C++


// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    int ans = 0;
public:
    int pathSum(TreeNode* root, int targetSum) {
        if (root) {
            paths(root, targetSum);
            pathSum(root->left, targetSum);
            pathSum(root->right, targetSum);
        }
        return ans;
    }
    void paths(TreeNode* root, long int target) {
        if (!root) return;
        if (root->val == target) ans++;
        paths(root->left, target - root->val);
        paths(root->right, target - root->val);
    }
};
